server/src/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -2639,22 +2639,23 @@ private int getMaxDataVolumesSupported(UserVmVO vm) {
         return maxDataVolumesSupported.intValue();
     }
 
-    private Long getDeviceId(long vmId, Long deviceId) {
+    private Long getDeviceId(UserVmVO vm, Long deviceId) {
         // allocate deviceId
-        List<VolumeVO> vols = _volsDao.findByInstance(vmId);
+        int maxDataVolumesSupported = getMaxDataVolumesSupported(vm);
```

Comment:
@HrWiggles Thanks for the review. The max data volumes here is the actual hypervisor capability (which is posted in the db). The device id 3 is being reserved for something since long and I don't want that to be effected.
When _getMaxDataVolumesSupported()_ returns 6, max 5 volumes can be attached to the VM and one device reserved (might be for virtual tools/CDROM). _maxDataVolumesSupported_ specifies the data volumes limit supported by hypervisor, nothing related to _maxDeviceId_.